# Лабораторная работа №3

Выполнил студент группы 9381 Никита Шахин.

## Цель работы: 
Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

## Задачи лабораторной работы
* Решить свой вариант диофантова уравнения.
* Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.
* Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.

**Вариант 17**

### 1. Решение диофантова уравнения.

$2023x + 4420y = 68$

In [4]:
from sympy.solvers.diophantine import diophantine
from sympy import symbols
from sympy.crypto.crypto import rsa_private_key, decipher_rsa
from sympy.logic.boolalg import to_cnf, to_dnf

x,y,z = symbols('x y z')

In [5]:
diophantine(2023*x + 4420*y - 68)

{(260*t_0 + 236, -119*t_0 - 108)}

Ответ: $x = 236,\; y = -108$

### 2.  Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.

Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом $(e = 7,m = 34)$.
Шифрованное сообщение имеет вид $(31; 26; 21; 14)$. Подберите закрытую часть ключа и прочитайте исходное слово.
Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “ё” и “ъ”).

In [6]:
code = [31, 26, 21, 14]
key = rsa_private_key(34,7) 
for i in code:
    print(decipher_rsa(i, key), end=" ")    

23 2 13 6 

Х А Л Д

## 3. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.

$f(x,\; y,\; z) = ((xy)\; xor\; (x\; xor\; z))\; ∨\; ((x\; ∨\; z)y)$

In [7]:
print("ДНФ: ", to_dnf(((x&y)^(x^z))|((x|z)&y), simplify=True))
print("КНФ: ", to_cnf(((x&y)^(x^z))|((x|z)&y), simplify=True))

ДНФ:  (x & y) | (x & ~z) | (z & ~x)
КНФ:  (x | z) & (y | ~x | ~z)


## Вывод

При выполнении лабораторной работы были изучены функции библиотеки SymPy Языка Программирования Python для решения задач дискретной математики и математической логики.